xla/service/gpu/cudnn_fused_mha_rewriter.cc

```diff
@@ -1205,38 +1326,66 @@ StatusOr<HloInstruction*> FuseFwdMultiHeadedAttentionBlock(
     algorithm->set_is_cudnn_frontend(true);
     algorithm->mutable_workspace_size()->set_value(0);
   }
+
+  // set is flash attention here
+  // choose to use flash attention or non-fa attention based on this flag.
+  fmha_config.set_is_flash_attention(is_flash_attention);
+  // set is_causal_mask here
+  // choose to generate causal mask inside cuDNN attention or not
+  fmha_config.set_is_causal_mask(is_causal_mask);
+
+  // Output Order: {O, scratch, Fwd act*}
   const Shape& output_shape = bmm_2->shape();
 
   Shape call_shape;
   // Activation output is used by backward gemm.
   HloInstruction* activation_output = nullptr;
 
-  std::vector<Shape> output_shapes = {output_shape,
-                                      ShapeUtil::MakeShape(U8, {0})};
+  std::vector<Shape> output_shapes = {
+      output_shape,
+      ShapeUtil::MakeShape(
+          U8, {is_flash_attention
+                   ? 16
+                   : 0})};  // reserved 2 int64 for dropout seed and offset
   if (is_training) {
-    // TODO Flush attention will have a different shape in training.
     activation_output = bmm_2->mutable_operand(0);
     // Sometimes activation output is bitcast, the actual activation is the
     // second user of the producer of bmm_2's first operand.
     if (activation_output->user_count() < 2 &&
         activation_output->opcode() == HloOpcode::kBitcast) {
       HloInstruction* producer = activation_output->mutable_operand(0);
       TF_RET_CHECK(producer->user_count() == 2);
-      activation_output = producer->UserId(activation_output) == 0
-                              ? producer->users()[1]
-                              : producer->users()[0];
+      HloInstruction* bmm2_grad2_user = producer->UserId(activation_output) == 0
```

Comment:
I think instead of checking producer->UserId you could also check producer->users()[0] == activation_output.
At least I think that is a little bit easier to understand.